### PR TITLE
fix: Handle kernel creation failure and destroy failed kernel

### PR DESCRIPTION
**Pull request by fregataa**
fix: #555

### CODE REVIEW

File: src/ai/backend/agent/agent.py [hunk 1]
```diff
@@ -865,6 +854,17 @@ async def _handle_destroy_event(self, ev: ContainerLifecycleEvent) -> None:
                     if ev.done_future is not None:
                         ev.done_future.set_exception(e)
                     raise
```
Comment (achimnol) [inline src/ai/backend/agent/agent.py:858]: What happens if the control goes here?
I think we should put the moved block inside a `finally` block following the above `try` block.

File: src/ai/backend/agent/docker/agent.py [hunk 2]
```diff
@@ -762,6 +762,7 @@ async def start_container(
                     config=container_config, name=kernel_name
                 )
                 cid = container._id
+                raise RuntimeError
```
Comment (achimnol) [inline src/ai/backend/agent/docker/agent.py]: This should be removed! 😉

File: src/ai/backend/agent/docker/agent.py [hunk 3]
```diff
@@ -778,9 +779,13 @@ async def start_container(
                             await writer.write(f"{k}={v}\n")
 
                 await container.start()
-            except asyncio.CancelledError:
-                raise
-            except Exception:
+            except asyncio.CancelledError as e:
+                try:
+                    cid = container._id
+                except (UnboundLocalError, AttributeError):
```
Comment (achimnol) [inline src/ai/backend/agent/docker/agent.py]: Could we remove this new inner `try` block?
I think this is the issue of control flow, so we could initialize `container` as `None` before entering the topmost `try` block and explicitly check if it's `None` here.

File: src/ai/backend/agent/docker/agent.py [hunk 4]
```diff
@@ -794,7 +799,11 @@ async def start_container(
                 async with self.resource_lock:
                     for dev_name, device_alloc in resource_spec.allocations.items():
                         self.computers[dev_name].alloc_map.free(device_alloc)
-                raise
+                try:
+                    cid = container._id
+                except (UnboundLocalError, AttributeError):
```
Comment (achimnol) [inline src/ai/backend/agent/docker/agent.py]: Same here.

**Review by achimnol — CHANGES_REQUESTED** [review]

File: src/ai/backend/agent/agent.py (hunk 1 above)
Comment (fregataa) [reply src/ai/backend/agent/agent.py:858]: I see.
I did not know that this flow will not put the `CLEAN` event in the queue.

File: src/ai/backend/agent/agent.py [hunk 5]
```diff
@@ -112,7 +112,7 @@
 
 from . import __version__ as VERSION
 from . import resources as resources_mod
-from .exception import AgentError, ResourceError
+from .exception import AgentError, DockerContainerCreationError, ResourceError
```
Comment (kyujin-cho) [inline src/ai/backend/agent/agent.py]: `DockerContainerCreationError`, name of the error, is too specific for docker backend. How about changing it to `ContainerCreationError`, without Docker?
```suggestion
from .exception import AgentError, ContainerCreationError, ResourceError
```

**Review by kyujin-cho — CHANGES_REQUESTED** [review]

File: src/ai/backend/agent/agent.py (hunk 5 above)
Comment (fregataa) [reply src/ai/backend/agent/agent.py]: I renamed it. Thanks!

File: src/ai/backend/agent/agent.py [hunk 6]
```diff
@@ -1590,12 +1592,28 @@ async def create_kernel(
...
+                preopen_ports,
+            )
+        except ContainerCreationError as e:
+            log.warning(
+                "Kernel failed to create container (k:{}). Kernel is going to be destroyed.",
+                ctx.kernel_id,
+            )
+            cid = e.container_id
+            async with self.registry_lock:
+                self.kernel_registry[ctx.kernel_id]["container_id"] = cid
```
Comment (kyujin-cho) [inline src/ai/backend/agent/agent.py:1644]: Please add `container_id` to the `AbstractKernel` definition.

File: src/ai/backend/agent/exception.py [hunk 7]
```diff
@@ -34,6 +34,12 @@ class UnsupportedBaseDistroError(RuntimeError):
     pass
 
 
+class ContainerCreationError(Exception):
+    def __init__(self, *args, container_id, **kwargs):
```
Comment (kyujin-cho) [inline src/ai/backend/agent/exception.py]: `*args` should appear before positional argument `container_id`.
```suggestion
    def __init__(self, container_id, *args, **kwargs):
```

Comment (fregataa) [reply src/ai/backend/agent/exception.py]: I locate `container_id` after positional argument cause I intended to set `container_id` only by keyword argument. is it okay?

Comment (fregataa) [reply src/ai/backend/agent/exception.py]: I updated it to `def __init__(self, container_id, *args, **kwargs):`

**Pull request merged**